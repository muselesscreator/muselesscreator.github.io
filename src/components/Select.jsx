import React from 'react';
import { v4 as uuid } from 'uuid';

import FormControl from '@mui/material/FormControl';
import InputLabel from '@mui/material/InputLabel';
import MuiSelect from '@mui/material/Select';
import MenuItem from '@mui/material/MenuItem';

import classNames from 'classnames';

const Select = ({
  label,
  options,
  value,
  handleChange,
}) => {
  const id = uuid();
  return (
    <FormControl className="mc-select-root">
      <InputLabel
      className="mc-select-label"
      id={`mc-select-label-${id}`}
    >
      {label}
    </InputLabel>
      <MuiSelect
        labelId={`mc-select-label-${id}`}
        className="filter-select mc-select"
        id={`mc-select-${id}`}
        value={value}
        label="Select category"
        onChange={(e) => {
          return handleChange(e.target.value);
        }}
      >
        {options.map(({ value, label }) => (
          <MenuItem
            key={value}
            value={value}
            className="mc-select-item"
          >
            {label}
          </MenuItem>
        ))}
      </MuiSelect>
    </FormControl>
  );
};
export default Select;

