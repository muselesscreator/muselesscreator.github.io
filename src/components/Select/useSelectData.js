import { v4 as uuid } from 'uuid';

export const selectLabelId = (id) => `mc-select-label-${id}`;
export const selectInputId = (id) => `mc-select-${id}`;

const useSelectData = ({ handleChange }) => {
  const id = uuid();
  const labelId = selectLabelId(id)
  const inputId = selectInputId(id)
  return {
    labelId,
    inputId,
     onChange: (e) => handleChange(e.target.value),
  };
};

export default useSelectData;
