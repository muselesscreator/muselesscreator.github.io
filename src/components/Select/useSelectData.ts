import { v4 as uuid } from 'uuid';

export const selectLabelId = (id: string) => `mc-select-label-${id}`;
export const selectInputId = (id: string) => `mc-select-${id}`;

const useSelectData = ({ handleChange }: { handleChange: (val: string) => unknown }) => {
  const id = uuid();
  const labelId = selectLabelId(id)
  const inputId = selectInputId(id)
  return {
    labelId,
    inputId,
     onChange: (e: { target: { value: string } }) => handleChange(e.target.value),
  };
};

export default useSelectData;
