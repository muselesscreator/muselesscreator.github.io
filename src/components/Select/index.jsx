import React from 'react';

import FormControl from '@mui/material/FormControl';
import InputLabel from '@mui/material/InputLabel';
import MuiSelect from '@mui/material/Select';
import MenuItem from '@mui/material/MenuItem';

import useSelectData from './useSelectData';

const Select = ({
  label,
  options,
  value,
  handleChange,
  multiple,
}) => {
  const {
    labelId,
    inputId,
    onChange,
  } = useSelectData({ handleChange });
  return (
    <FormControl className="mc-select-root">
      <InputLabel
        className="mc-select-label"
        id={labelId}
      >
        {label}
      </InputLabel>
      <MuiSelect
        labelId={labelId}
        className="filter-select mc-select"
        id={inputId}
        value={value}
        label="Select category"
        multiple={multiple}
        onChange={onChange}
      >
        {options.map(({ value, label }) => (
          <MenuItem
            key={value}
            value={value}
            className="mc-select-item"
          >
            {label}
          </MenuItem>
        ))}
      </MuiSelect>
    </FormControl>
  );
};
export default Select;

