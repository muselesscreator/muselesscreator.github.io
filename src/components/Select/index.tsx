import 'react';
import FormControl from '@mui/material/FormControl';
import InputLabel from '@mui/material/InputLabel';
import MuiSelect from '@mui/material/Select';
import MenuItem from '@mui/material/MenuItem';

import useSelectData from './useSelectData';

type Props = {
  label: string;
  options: { value: string; label: string }[];
  value: string | undefined;
  handleChange: (value: string | string[]) => void;
  multiple?: boolean;
}

const Select = ({
  label,
  options,
  value,
  handleChange,
  multiple,
}: Props) => {
  const {
    labelId,
    inputId,
    onChange,
  } = useSelectData({ handleChange });
  return (
    <FormControl className="mc-select-root">
      <InputLabel
        className="mc-select-label"
        id={labelId}
      >
        {label}
      </InputLabel>
      <MuiSelect
        labelId={labelId}
        className="filter-select mc-select"
        id={inputId}
        value={value}
        label="Select category"
        multiple={multiple}
        onChange={onChange}
      >
        {options.map(({ value, label }) => (
          <MenuItem
            key={value}
            value={value}
            className="mc-select-item"
          >
            {label}
          </MenuItem>
        ))}
      </MuiSelect>
    </FormControl>
  );
};
export default Select;

