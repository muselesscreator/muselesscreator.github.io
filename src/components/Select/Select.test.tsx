import 'react';
import {
  beforeEach,
  describe,
  test,
  expect,
  vi,
  Mock,
} from 'vitest';
import {
  shallow,
  setMockName,
  ExplorerData,
} from '@muselesscreator/react-shallow-snapshot';

import Select from '.';
import useSelectData from './useSelectData';

vi.mock('@mui/material/FormControl', () => ({ default: 'FormControl' }));
vi.mock('@mui/material/InputLabel', () => ({ default: 'InputLabel' }));
vi.mock('@mui/material/Select', () => ({ default: 'MuiSelect' }));
vi.mock('@mui/material/MenuItem', () => ({ default: 'MenuItem' }));

vi.mock('uuid', () => ({ v4: () => 'test-uuid' }));
vi.mock('./useSelectData', () => ({ default: vi.fn() }));


const props = {
  label: 'test-label',
  value: 'test-value-1',
  options: [
    { value: 'test-value1', label: 'Test label1' },
    { value: 'test-value2', label: 'Test label2' },
    { value: 'test-value3', label: 'Test label3' },
  ],
  handleChange: vi.fn(),
  multiple: true,
}
const hookProps = {
  labelId: 'test-label-id',
  inputId: 'test-input-id',
  onChange: vi.fn(),
};
(useSelectData as Mock).mockReturnValue(hookProps);
setMockName(props.handleChange, 'props.handleChange');
setMockName(hookProps.onChange, 'hookProps.onChange');
describe('Select component', () => {
  let instance;
  let els = {}  as Record<string, typeof ExplorerData>;
  beforeEach(() => {
    vi.clearAllMocks();
  }); 
  describe('render', () => {
    test('snapshot', () => {
      const el = shallow(<Select {...props} />);
      instance = el.instance;
      expect(el.snapshot).toMatchSnapshot();
    });
    test('render form control with children', () => {
      const el = shallow(<Select {...props} />);
      instance = el.instance;
      els = {
        FormControl: instance.findByType('FormControl')[0],
        InputLabel: instance.findByType('InputLabel')[0],
      };
      expect(els.FormControl.children[0].matches(els.InputLabel)).toBe(true);
      expect(els.InputLabel.props.id).toEqual(hookProps.labelId);
      expect(els.InputLabel.children[0].el).toEqual(props.label);
      const selectEl = els.FormControl.children[1].el as typeof ExplorerData;
      expect(selectEl.type).toEqual('MuiSelect');
    });
    test('render select with props', () => {
      const el = shallow(<Select {...props} />);
      instance = el.instance;
      els = {
        FormControl: instance.findByType('FormControl')[0],
        MuiSelect: instance.findByType('MuiSelect')[0],
      };
      const selectProps = els.MuiSelect.props;
      expect(selectProps.labelId).toEqual(hookProps.labelId);
      expect(selectProps.id).toEqual(hookProps.inputId);
      expect(selectProps.onChange).toEqual(hookProps.onChange);
      expect(selectProps.value).toEqual(props.value);
      expect(selectProps.multiple).toEqual(props.multiple);
      expect(els.MuiSelect.children).toHaveLength(props.options.length);

      props.options.forEach((option, i) => {
        const selectEl = els.MuiSelect.children[i].el as typeof ExplorerData;
        expect(selectEl.type).toEqual('MenuItem');
        expect(els.MuiSelect.children[i].props.value).toEqual(option.value);
        expect(els.MuiSelect.children[i].children[0].el).toEqual(option.label);
      });
    });
  });
});
