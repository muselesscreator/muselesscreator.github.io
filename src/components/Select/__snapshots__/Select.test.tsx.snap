// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select component > render > snapshot 1`] = `
"<FormControl
  className="mc-select-root"
>
  <InputLabel
    className="mc-select-label"
    id="test-label-id"
  >
    test-label
  </InputLabel>
  <MuiSelect
    className="filter-select mc-select"
    id="test-input-id"
    label="Select category"
    labelId="test-label-id"
    multiple="test-multiple"
    onChange={[Function hookProps.onChange]}
    value="test-value-1"
  >
    <MenuItem
      className="mc-select-item"
      value="test-value1"
    >
      Test label1
    </MenuItem>
    <MenuItem
      className="mc-select-item"
      value="test-value2"
    >
      Test label2
    </MenuItem>
    <MenuItem
      className="mc-select-item"
      value="test-value3"
    >
      Test label3
    </MenuItem>
  </MuiSelect>
</FormControl>"
`;
